network/src/noise/handshake.rs

```diff
@@ -726,15 +788,257 @@ mod test {
+        let server_peer_id = server.network_context.peer_id();
+        let server_network_context =
+            NetworkContext::new(RoleType::Validator, NetworkId::Vfn, server_peer_id);
+        server.network_context = server_network_context;
+
+        // Add the validator to the trusted peers set
+        let trusted_peers = peers_and_metadata
+            .get_trusted_peers(&NetworkId::Vfn)
+            .unwrap();
+        trusted_peers.write().insert(
+            server_peer_id,
+            Peer::new(
+                vec![],
+                [server_public_key].into_iter().collect(),
+                PeerRole::Validator,
+            ),
+        );
+
+        // Create an in-memory socket for testing
+        let (dialer_socket, listener_socket) = MemorySocket::new_pair();
+
+        // Create the client connection task
+        let client_connection_task = async move {
+            let (_, peer_role) = client
+                .upgrade_outbound(
+                    dialer_socket,
+                    server_peer_id,
+                    server_public_key,
+                    AntiReplayTimestamps::now,
+                )
+                .await
+                .unwrap();
+            assert_eq!(peer_role, PeerRole::Validator);
+        };
+
+        // Create the server connection task
+        let server_connection_task = async move {
+            let (_, peer_id, peer_role) = server.upgrade_inbound(listener_socket).await.unwrap();
+            assert_eq!(peer_id, client_peer_id);
+            assert_eq!(peer_role, PeerRole::ValidatorFullNode);
+        };
+
+        // Perform the handshake
+        block_on(join(client_connection_task, server_connection_task));
+    }
+
+    #[test]
+    fn test_handshake_peer_roles_vfn_dials_vfn() {
+        // Initialize the logger
+        ::aptos_logger::Logger::init_for_testing();
+
+        // Create a peers and metadata struct with no trusted peers
+        let network_ids = vec![NetworkId::Vfn, NetworkId::Public];
+        let peers_and_metadata = PeersAndMetadata::new(&network_ids);
+
+        // Create a client and server with mutual auth disabled
+        let ((mut client, _), (mut server, server_public_key)) =
+            build_peers(false, Some(peers_and_metadata.clone()));
+
+        // Update the client network context
+        let client_peer_id = client.network_context.peer_id();
+        let client_network_context =
+            NetworkContext::new(RoleType::FullNode, NetworkId::Public, client_peer_id);
+        client.network_context = client_network_context;
+
+        // Update the server network context
+        let server_peer_id = server.network_context.peer_id();
+        let server_network_context =
+            NetworkContext::new(RoleType::FullNode, NetworkId::Public, server_peer_id);
+        server.network_context = server_network_context;
+
+        // Add the server VFN to the trusted peers set
+        let trusted_peers = peers_and_metadata
```

Comment:
I see that in this test we only add the server VFN to the trusted_peers.

In a "normal" network, do we expect both to be in trusted_peers, so they both identify each other as vfns?

Reply:
Aah, good point. I just double checked this and in a "normal" network, both are in the trusted set, so I've updated the test to ensure they both connect as VFNs 😄